File: src/main/java/com/zuminX/utils/builder/ApiImplicitParamGenerator.java
package com.zuminX.utils.builder;

import cn.hutool.core.bean.OptionalBean;
import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiIdentifier;
import com.intellij.psi.PsiModifierList;
import com.intellij.psi.PsiParameter;
import com.zuminX.annotations.swagger.ApiImplicitParam;
import com.zuminX.interceptor.AnnotationGeneratorInterceptor;
import com.zuminX.names.ClassName;
import com.zuminX.names.RequestAnnotation;
import com.zuminX.utils.PsiUtils;
import com.zuminX.utils.PublicUtils;
import java.util.Arrays;
import java.util.Objects;

/**
 * ApiImplicitParam注解的生成器类
 */
public class ApiImplicitParamGenerator implements AnnotationGenerator<PsiParameter, ApiImplicitParam> {

  public static final ApiImplicitParamGenerator INSTANCE = AnnotationGeneratorInterceptor.create(ApiImplicitParamGenerator.class, ApiImplicitParam.class);

  /**
   * 根据psi参数对象构建ApiImplicitParam注解对象
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam注解对象
   */
  public final ApiImplicitParam build(PsiParameter psiParameter) {
    return ApiImplicitParam.builder()
        .paramType(getParamType(psiParameter))
        .dataType(getDataType(psiParameter))
        .dataTypeClass(getDataTypeClass(psiParameter))
        .name(getName(psiParameter))
        .value(getValue(psiParameter))
        .build();
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的name属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的name属性
   */
  protected String getName(PsiParameter psiParameter) {
    return OptionalBean.ofNullable(psiParameter.getNameIdentifier()).getBean(PsiIdentifier::getText).get();
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的dataType属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的dataType属性
   */
  protected String getDataType(PsiParameter psiParameter) {
    return PublicUtils.getSimpleNameByQualifiedName(getParameterType(psiParameter));
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的dataTypeClass属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的dataTypeClass属性
   */
  protected ClassName getDataTypeClass(PsiParameter psiParameter) {
    return new ClassName(getParameterType(psiParameter));
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的paramType属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的paramType属性
   */
  protected String getParamType(PsiParameter psiParameter) {
    PsiModifierList modifierList = psiParameter.getModifierList();
    if (modifierList == null) {
      return null;
    }
    return Arrays.stream(modifierList.getAnnotations())
        .map(PsiAnnotation::getQualifiedName)
        .map(RequestAnnotation::findByQualifiedName)
        .filter(Objects::nonNull)
        .findAny()
        .map(RequestAnnotation::getType)
        .orElse(null);
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的value属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的value属性
   */
  protected String getValue(PsiParameter psiParameter) {
    return PsiUtils.getFirstComment(psiParameter);
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的parameterType属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的parameterType属性
   */
  private String getParameterType(PsiParameter psiParameter) {
    return psiParameter.getType().getCanonicalText();
  }

}

File: src/main/java/com/zuminX/generator/builder/ApiImplicitParamGenerator.kt
package com.zuminX.generator.builder

import com.intellij.psi.PsiParameter
import com.zuminX.annotations.swagger.ApiImplicitParam
import com.zuminX.generator.AnnotationGenerator
import com.zuminX.generator.AnnotationGeneratorInterceptor
import com.zuminX.names.ClassName
import com.zuminX.names.annotation.RequestAnnotation
import com.zuminX.utils.getFirstComment
import com.zuminX.utils.getSimpleName

val ApiImplicitParamGeneratorInstance =
  AnnotationGeneratorInterceptor.create(ApiImplicitParamGenerator::class.java, ApiImplicitParam::class.java)

/**
 * ApiImplicitParam注解的生成器类
 */
open class ApiImplicitParamGenerator :
  AnnotationGenerator<PsiParameter, ApiImplicitParam> {

  /**
   * 根据psi参数对象构建ApiImplicitParam注解对象
   *
   * @param psiElement psi参数
   * @return ApiImplicitParam注解对象
   */
  final override fun build(psiElement: PsiParameter): ApiImplicitParam {
    return ApiImplicitParam().apply {
      paramType = getParamType(psiElement)
      dataType = getDataType(psiElement)
      dataTypeClass = getDataTypeClass(psiElement)
      name = getName(psiElement)
      value = getValue(psiElement)
    }
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的name属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的name属性
   */
  protected fun getName(psiParameter: PsiParameter): String? {
    return psiParameter.nameIdentifier?.text
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的dataType属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的dataType属性
   */
  protected fun getDataType(psiParameter: PsiParameter): String {
    return getParameterType(psiParameter).getSimpleName()
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的dataTypeClass属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的dataTypeClass属性
   */
  protected fun getDataTypeClass(psiParameter: PsiParameter): ClassName {
    return ClassName(getParameterType(psiParameter))
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的paramType属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的paramType属性
   */
  protected fun getParamType(psiParameter: PsiParameter): String? {
    val modifierList = psiParameter.modifierList ?: return null
    return modifierList.annotations
      .map { it.qualifiedName }
      .mapNotNull { RequestAnnotation.findByQualifiedName(it) }
      .map { it.type }
      .firstOrNull()
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的value属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的value属性
   */
  protected fun getValue(psiParameter: PsiParameter): String? {
    return psiParameter.getFirstComment()
  }

  /**
   * 从psi参数对象中获取ApiImplicitParam类的parameterType属性
   *
   * @param psiParameter psi参数
   * @return ApiImplicitParam类的parameterType属性
   */
  private fun getParameterType(psiParameter: PsiParameter): String {
    return psiParameter.type.canonicalText
  }
}